e2e-test/e2e.py

```diff
@@ -97,6 +102,166 @@ def test_post_resource_increase_count(
+    """Test to add a resource to the backend via the Proxy and verify AuditEvent creation."""
+    _test_post_and_verify_audit_events(
+        "Observation",
+        file_name,
+        1,
+        hapi,
+        fhir_proxy,
+        auth
+    )
+
+
+@patch.dict(os.environ, {"AUDIT_EVENT_LOGGING_ENABLED_ENV": "true"})
+def test_post_bundle_with_logging_enabled_creates_audit_events(
+    file_name: str,
+    hapi: clients.HapiClient,
+    fhir_proxy: clients.FhirProxyClient,
+    auth: clients.AuthClient,
+) -> None:
+    """Test to add a bundle to the backend via the Proxy and verify AuditEvents creation."""
+    _test_post_and_verify_audit_events(
+        "",
+        file_name,
+        2,
+        hapi,
+        fhir_proxy,
+        auth
+    )
+
+ 
+def _test_post_and_verify_audit_events(
+    resource_type: str,
+    file_name: str,
+    expected_audit_event_increase: int,
+    hapi: clients.HapiClient,
+    fhir_proxy: clients.FhirProxyClient,
+    auth: clients.AuthClient
+) -> None:
+    """Helper function to POST a resource and verify AuditEvent creation."""
+    token = auth.get_auth_token()
+    payload_type = "Bundle" if resource_type == "" else resource_type
+
+    initial_audit_event_count = hapi.get_audit_event_count()
+    logging.info("Initial AuditEvent count: %d", initial_audit_event_count)
+
+    fhir_proxy.post_resource(resource_type, file_name, token)
+    logging.info("Posted %s resource via Proxy.", payload_type)
+
+    max_wait_seconds = 300
+    poll_interval = 5
+    waited = 0
+    while waited < max_wait_seconds:
+        current_audit_events = hapi.get_audit_events(expected_audit_event_increase)
+        current_audit_event_count = hapi.get_audit_event_count()
+        if current_audit_event_count == initial_audit_event_count + expected_audit_event_increase:
+            expected_output_filename = ("transaction_bundle_audit_events.json" 
+            if resource_type == "" 
+            else "{resource_type}_audit_events.json".format(resource_type=resource_type)
+            ).lower()                           
+            for audit_event_entry in current_audit_events:
+                audit_event = cast(Dict[str, str], audit_event_entry.get("resource", {}))
+                _assert_audit_events(read_file("e2e-test/{}".format(expected_output_filename)), audit_event)
+                    
+            logging.info("AuditEvents created successfully after %s POST.", payload_type)
+            return
+        time.sleep(poll_interval)
+        waited += poll_interval
+        logging.info("POST %s :: Waiting for AuditEvent(s)... (%ds elapsed)", payload_type, waited)
+
+    raise AssertionError(
+        "Valid AuditEvent was NOT created after {} POST within {} seconds. "
+        "Initial count: {}, Current count: {}".format(
+            payload_type, max_wait_seconds, initial_audit_event_count, current_audit_event_count
+        )
+    )   
+
+def _assert_audit_events(expected_audit_event: Dict[str, str],
```

Comment:
Here and everywhere else that you are casting JSON objects to `Dict`, I am wondering if this is required? I mean can't you directly work with JSON objects, e.g., if `audit_event = audit_event_entry.get("resource", {})` I think you should be able to directly access its keys, e.g., `audit_event["agent"]`, no?